### ISSUE TITLE
Installation and usability issues Oct 2022

### ISSUE THREAD

**Issue by agitter** [gitter-lab/ssps#27]
@evangorstein reported that the conda instructions in the readme give an UnsatisfiableError at the line `conda install -c bioconda -c conda-forge snakemake`.  I was not able to reproduce this on my system, but I am not surprised because the current instructions' behavior could be quite dependent on how Python and Anaconda are configured locally.

There are a few ways to improve our current setup:
1) Installing all dependencies in one step so that conda can avoid conflicts
2) Pinning versions of essential packages
3) Switching to snakemake-minimal instead of snakemake

For 1), I propose this new command that would create a new environment and install the required packages all at once:
```
conda create -n ssps -c conda-forge pandas matplotlib numpy bioconda::snakemake-minimal
```
@evangorstein does that work on your system?  If it does, I'll update the readme.

For 2), I don't think we need to make updates now.  The conda environment is primarily used to run the Snakemake workflow and generate figures.  It should not be too dependent on specific package versions.  In case we need to pin versions later, here are the versions known to work together and work with our code in my local environment:
<details>
<summary>conda list output</summary>

```
# Name                    Version                   Build  Channel
appdirs                   1.4.3                      py_1    conda-forge
attrs                     19.3.0                     py_0    conda-forge
brotlipy                  0.7.0           py37h4ab8f01_1000    conda-forge
ca-certificates           2020.4.5.1           hecc5488_0    conda-forge
certifi                   2020.4.5.1       py37hc8dfbb8_0    conda-forge
cffi                      1.14.0           py37ha419a9e_0    conda-forge
chardet                   3.0.4           py37hc8dfbb8_1006    conda-forge
configargparse            1.2.3              pyh9f0ad1d_0    conda-forge
cryptography              2.9.2            py37h26f1ce3_0    conda-forge
datrie                    0.8.2            py37h8055547_0    conda-forge
decorator                 4.4.2                      py_0    conda-forge
docutils                  0.16             py37hc8dfbb8_1    conda-forge
gitdb                     4.0.5                      py_0    conda-forge
gitpython                 3.1.2                      py_0    conda-forge
idna                      2.9                        py_1    conda-forge
importlib-metadata        1.6.0            py37hc8dfbb8_0    conda-forge
importlib_metadata        1.6.0                         0    conda-forge
intel-openmp              2020.0                      166
ipython_genutils          0.2.0                      py_1    conda-forge
jsonschema                3.2.0            py37hc8dfbb8_1    conda-forge
jupyter_core              4.6.3            py37hc8dfbb8_1    conda-forge
libblas                   3.8.0                    15_mkl    conda-forge
libcblas                  3.8.0                    15_mkl    conda-forge
liblapack                 3.8.0                    15_mkl    conda-forge
mkl                       2020.0                      166
nbformat                  5.0.6                      py_0    conda-forge
numpy                     1.18.4           py37hae9e721_0    conda-forge
openssl                   1.1.1g               he774522_0    conda-forge
pandas                    1.0.3            py37h3bbf574_1    conda-forge
pip                       20.1               pyh9f0ad1d_0    conda-forge
psutil                    5.7.0            py37h8055547_1    conda-forge
pycparser                 2.20                       py_0    conda-forge
pyopenssl                 19.1.0                     py_1    conda-forge
pyrsistent                0.16.0           py37h8055547_0    conda-forge
pysocks                   1.7.1            py37hc8dfbb8_1    conda-forge
python                    3.7.6           h60c2a47_5_cpython    conda-forge
python-dateutil           2.8.1                      py_0    conda-forge
python_abi                3.7                     1_cp37m    conda-forge
pytz                      2020.1             pyh9f0ad1d_0    conda-forge
pywin32                   227              py37hfa6e2cd_0    conda-forge
pyyaml                    5.3.1            py37h8055547_0    conda-forge
ratelimiter               1.2.0                 py37_1000    conda-forge
requests                  2.23.0             pyh8c360ce_2    conda-forge
setuptools                46.3.1           py37hc8dfbb8_0    conda-forge
six                       1.14.0                     py_1    conda-forge
smmap                     3.0.4              pyh9f0ad1d_0    conda-forge
snakemake-minimal         5.17.0                     py_0    bioconda
sqlite                    3.30.1               hfa6e2cd_0    conda-forge
toposort                  1.5                        py_3    conda-forge
traitlets                 4.3.3            py37hc8dfbb8_1    conda-forge
urllib3                   1.25.9                     py_0    conda-forge
vc                        14.1                 h869be7e_1    conda-forge
vs2015_runtime            14.16.27012          h30e32a0_2    conda-forge
wheel                     0.34.2                     py_1    conda-forge
win_inet_pton             1.1.0                    py37_0    conda-forge
wincertstore              0.2                   py37_1003    conda-forge
wrapt                     1.12.1           py37h8055547_1    conda-forge
yaml                      0.2.4                he774522_0    conda-forge
zipp                      3.1.0                      py_0    conda-forge
```
</details>

For 3), I suggest we switch to snakemake-minimal to avoid the large number of dependencies that are required to support cloud computing with Snakemake.  We can make installing the snakemake package an optional second step along with cookiecutter.

**Comment by dpmerrell** [comment]
Thanks for raising this @agitter, it's been a couple of years since I last updated this codebase -- I expect some things to be broken.

I just tried running your proposed changes on my Ubuntu 18 workstation. The commands work for me, I'm happy to approve the update to README.

Unfortunately I ran into Julia issues when I actually ran the workflow -- some dependencies have made breaking API changes over the past couple of years. 

Bottom line, this codebase needs some attention, I'll devote some time to updates over the coming days.

**Comment by evangorstein** [comment]
@agitter, your command that create a new environment and installs the required packages all at once worked on my laptop. Thanks! @dpmerrell, I also run into a lot of errors when I run the analysis with `snakemake --cores 8`. I have Julia 1.8 and also Julia 1.4 installed on my computer. I wonder if snakemake know to use Julia 1.4 when the workflow is run? I'm very much over my head with this, so I apologize if this is a dumb question.

**Comment by dpmerrell** [comment]
@evangorstein I have bad news for you -- you'll also need to install Julia 1.6 now :)

To answer your question: no, Snakemake will not automatically choose the right version of Julia.
Its rules simply call `julia` from the shell. 
One workaround is to execute `alias julia=PATH/TO/JULIA1.6/BINARY` at the terminal before running the Snakemake workflow.

**Comment by dpmerrell** [comment]
@evangorstein To clarify my above cryptic comment: this afternoon I updated (A) the Julia version to 1.6 and (B) all of the package dependencies.

So I recommend running `git pull` to get the most current version of the code, and stepping through the Julia instructions from README again.

Thanks for your interest in this codebase, I'm happy to answer any questions you have :)

**Comment by evangorstein** [comment]
Thanks so much for your help. When I run the analysis, I'm now running into an error message that one of my Julia packages is not installed correctly: 
<img width="1021" alt="image" src="https://user-images.githubusercontent.com/22649919/197416833-c41d2d3a-b77d-4d44-b2dc-ee5a3b30d790.png">
I've instantiated the Manifest and Project.toml files in the folder `SSPS`, and I run `alias julia="/Applications/Julia-1.6.app/Contents/Resources/julia/bin/julia --project=SSPS"` prior to calling snakemake.

I think `mclust` might also be a required R package that needs to be installed before running the workflow if you want to also mention that in the Readme.

**Comment by dpmerrell** [comment]
Hi Evan,

Some responses to your comments/questions:

1. Looking at your screenshot, I see the printout mention Julia 1.8 (rather than 1.6). So part of the issue is probably that the wrong version of Julia is still being called.
2. Including the `--project` flag in your alias is most likely redundant, since the Snakemake rules already specify `--project`. For example, it appears [in this line of the Snakefile.](https://github.com/gitter-lab/ssps/blob/be84c7c4f988b39e0cabd20cdf68bf9620c35f0d/Snakefile#L174)
3. I'll look into the `mclust` dependency -- thanks for bringing it up.

Note: For documentation purposes, I've renamed this issue to catch a broader class of installation/usability problems.

**Comment by evangorstein** [comment]
Yeah, I'm not sure why it's still calling the wrong version of Julia. Do you have any suggestions for a workaround?

**Comment by dpmerrell** [comment]
One thing to be aware of: an `alias` only applies to your current terminal session; it doesn't apply to other terminals you open or if you close/reopen the terminal.

If that doesn't catch it, then I'm not sure what the issue might be.

Also @evangorstein feel free to send me an email (dmerrell at cs dot wisc dot edu) if you want us to troubleshoot this in-person. 

I know it can be challenging to simultaneously learn Julia, Snakemake, Anaconda, Python, bash, etc. :)

**Comment by evangorstein** [comment]
@dpmerrell, thanks for your help. Yeah, that would be great--I'll shoot you an email! :)

**Comment by dpmerrell** [comment]
Fixed the problem in person -- it resulted from local environment. 
I'll close this issue, but feel free to reach out again if you encounter problems @evangorstein
